// @glennsl/bs-json/src/Json.js
// Generated by Melange

import * as Js_exn from "melange.belt/./js_exn.js";
import * as Caml_option from "melange.runtime/caml_option.js";
import * as Caml_exceptions from "melange.runtime/caml_exceptions.js";
import * as Caml_js_exceptions from "melange.runtime/caml_js_exceptions.js";

var ParseError = /* @__PURE__ */Caml_exceptions.create("Json.ParseError");

function parse(s) {
  try {
    return Caml_option.some(JSON.parse(s));
  }
  catch (exn){
    return ;
  }
}

function parseOrRaise(s) {
  try {
    return JSON.parse(s);
  }
  catch (raw_e){
    var e = Caml_js_exceptions.internalToOCamlException(raw_e);
    if (e.RE_EXN_ID === Js_exn.$$Error) {
      var m = e._1.message;
      var message = m !== undefined ? m : "Unknown error";
      throw {
            RE_EXN_ID: ParseError,
            _1: message,
            Error: new Error()
          };
    }
    throw e;
  }
}

function stringify(prim) {
  return JSON.stringify(prim);
}

export {
  ParseError ,
  parse ,
  parseOrRaise ,
  stringify ,
}
/* No side effect */
