// @glennsl/bs-json/src/Json.js
// Generated by Melange
'use strict';

var Js_exn = require("melange.belt/./js_exn.js");
var Caml_option = require("melange.runtime/caml_option.js");
var Caml_exceptions = require("melange.runtime/caml_exceptions.js");
var Caml_js_exceptions = require("melange.runtime/caml_js_exceptions.js");

var ParseError = /* @__PURE__ */Caml_exceptions.create("Json.ParseError");

function parse(s) {
  try {
    return Caml_option.some(JSON.parse(s));
  }
  catch (exn){
    return ;
  }
}

function parseOrRaise(s) {
  try {
    return JSON.parse(s);
  }
  catch (raw_e){
    var e = Caml_js_exceptions.internalToOCamlException(raw_e);
    if (e.RE_EXN_ID === Js_exn.$$Error) {
      var m = e._1.message;
      var message = m !== undefined ? m : "Unknown error";
      throw {
            RE_EXN_ID: ParseError,
            _1: message,
            Error: new Error()
          };
    }
    throw e;
  }
}

function stringify(prim) {
  return JSON.stringify(prim);
}

exports.ParseError = ParseError;
exports.parse = parse;
exports.parseOrRaise = parseOrRaise;
exports.stringify = stringify;
/* No side effect */
